zio/bzngio/index.go

```diff
@@ -0,0 +1,183 @@
+package bzngio
+
+import (
+	"errors"
+	"io"
+	"os"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type Index struct {
+	index      []mark
+	indexReady bool
+}
+
+type IndexReader interface {
+	zbuf.ReadCloser
+	// count of how many records were read from disk, currently just
+	// used for testing.
+	Reads() uint64
+}
+
+type mark struct {
+	Ts     nano.Ts
+	Offset int64
+}
+
+// NewIndex creates a new Index object, which is the container that holds
+// the in-memory index for a (b)zng file.  The first call to NewReader()
+// will return a reader that scans the entire file, building a time-based
+// index in the process, subsequent readers can use this index to read
+// only the relevant zng streams from the underlying file.
+func NewIndex() Index {
+	return Index{}
+}
+
+func (i *Index) NewReader(f *os.File, zctx *resolver.Context, span nano.Span) (IndexReader, error) {
+	if i.indexReady {
+		return newRangeReader(f, zctx, i.index, span)
+	} else {
+		return &indexReader{
+			Reader: *NewReader(f, zctx),
+			closer: f,
+			start:  span.Ts,
+			end:    span.End(),
+			parent: i,
+		}, nil
+	}
+}
+
+// indexReader is a zbuf.Reader that also builds an index as it reads.
+type indexReader struct {
+	Reader
+	closer  io.Closer
+	start   nano.Ts
+	end     nano.Ts
+	parent  *Index
+	marks   []mark
+	lastSOS int64
+	lastTs  nano.Ts
+}
+
+func (i *indexReader) Read() (*zng.Record, error) {
+	for {
+		rec, err := i.readOne()
+		if rec == nil {
+			i.parent.index = i.marks
+			i.parent.indexReady = true
+			return nil, nil
+		}
+
+		if err != nil {
+			return nil, err
+		}
```

Comment:
This should probably happen before the `if rec == nil` statement.

Reply:
Should it?  The contract for Read() isn't spelled out anywhere but if it returns (nil, io.EOF) then this check needs to come first (or move inside the err check)

Reply:
It should. See #587.